### DISCUSSION TITLE
Anyone ever use ElSql with Jdbi SQL Objects???

### DISCUSSION THREAD

**Discussion by mattbertolini** [jdbi/jdbi#2695] [Q&A] [answered]
This is probably me missing something pretty obvious, but has anyone successfully integrated [ElSql](https://github.com/OpenGamma/ElSql) with Jdbi SQL objects API? I've used it successfully with the core API as it is relatively imperative and straightforward, but I've always wanted to use it in the SQL objects API via some annotations. 

The first approach I tried was to create a `SqlLocator`. It felt like the natural location as the ElSql API is a bit convention-based and locates its files via classpath convention. But ElSql also requires the bound Sql parameters at the same time and that data is not available in the `SqlLocator` interface. In theory, I could pass an empty `Map` for the SQL params, but that will cause an exception with ElSql if certain (and quite useful) features are utilized.

```java
public class ElSqlSqlLocator implements SqlLocator {
    @Override
    public String locate(Class<?> sqlObjectType, Method method, ConfigRegistry config) {
        final ElSql elSql = ElSql.of(ElSqlConfig.DEFAULT, sqlObjectType);
        return elSql.getSql(method.getName(), sqlParams); // Don't have SQL params here???
    }
}
```

Next I thought to create a `TemplateEngine` since ElSql is a bit of a template language. I have the `StatementContext` in the template engine, but I'm not sure I have enough information about the SQL object type to fill out the necessary arguments for ElSql.

I've also considered a `StatementCustomizer` but the same issues as the `TemplateEngine` come up. Am I missing something or is this the ElSql API just not going to fit into the SQL objects API? I would appreciate any advice you all have. Many thanks!

**Comment by stevenschlansker** [comment]
Hi @mattbertolini , thanks for your interest. I've never heard of ElSql before myself, nor heard of anyone else attempting this.
I think based on your description the closest interface is `TemplateEngine` which takes a sql template and renders final sql.
`SqlLocator` is for finding and loading sql templates (before rendering), and `StatementCustomizer` happens after the sql is rendered.

As you point out, there is then some confusion about who is responsible for transforming the `:parameter` named parameter style into JDBC which only supports `?` positional parameters. It sounds like both Jdbi and ElSql expect to handle this.

The parameters are not available at rendering time, since Jdbi is trying to prepare the statement to be executed potentially many times with varying parameters. The existing template engines today instead feed in the "attributes", which are more like traditional template string parameters, not JDBC bound parameters.

Would it work to feed the attributes in to ElSql to render a SQL statement that retains the `:parameter`, and feed that in to Jdbi? You can get the attributes from `config.get(SqlStatements.class).getAttributes()`

**Reply by mattbertolini** [reply]
I'm struggling to figure out how to get the necessary data to leverage a TemplateEngine.

An ElSql file typically is convention based. The name of the Elsql file maps loosely to a DAO class of some kind. Inside the file they have multiple queries with a given name. Those names are typically the method names inside of the DAO class. That maps really well over to Jdbi Sql objects.

ElSql requires three things:

* The Sql object type (Class<?>). This is to locate the ElSql file.
* The Sql object method name (Method). This is used to locate the query in the ElSql file.
* The set of parameters. This doesn't mean they will be bound to the :parameters but they can be used to determine some of what the ElSql DSL does.

I am unaware of how to get the first two things in a `TemplateEngine`. The engine interface seems to rely on the fact that the template is loaded as a String and part of ElSql's job is to locate the template as well. 

I can definitely get two of those things in a `SqlLocator`. I like the idea of using the attributes but I think I would need to pass around the `ElSql` type rather than something more primitive.

**Reply by stevenschlansker** [reply]
Yeah, in this case, I think you implement both a SqlLocator (to find the text) and a TemplateEngine (to apply ElSql to the loaded text).

Is it possible to separate the ElSql steps of loading the text, versus rendering it as a template? If not, this might be tricky.

**Comment by mattbertolini** [comment, marked as answer]
Many thanks for the info @stevenschlansker. Between this conversation and a careful reading of the StringTemplateEngine I think I was able to get a path forward.

I've created an extension configured that implements both a SqlLocator and TemplateEngine as you stated:

```java
public class ElSqlBundleSqlExtensionConfigurer extends SimpleExtensionConfigurer {
    @Override
    public void configure(ConfigRegistry config, Annotation annotation, Class<?> extensionType) {
        final ElSql elSql = ElSql.of(ElSqlConfig.DEFAULT, extensionType);
        config.get(SqlObjects.class).setSqlLocator(new ElSqlBundleSqlLocator());
        config.get(SqlStatements.class).setTemplateEngine(new ElSqlBundleTemplateEngine(elSql));
    }
}

public class ElSqlBundleSqlLocator implements SqlLocator {
    @Override
    public String locate(Class<?> sqlObjectType, Method method, ConfigRegistry config) {
        // TODO: Potentially check an annotation for a different name
        return method.getName();
    }
}

public class ElSqlBundleTemplateEngine implements TemplateEngine {
    private final ElSql elSqlBundle;

    public ElSqlBundleTemplateEngine(ElSql elSqlBundle) {
        this.elSqlBundle = elSqlBundle;
    }

    @Override
    public String render(String template, StatementContext ctx) {
        // The "template" in this case is the name returned from the SqlLocator
        final Map<String, Object> attributes = ctx.getConfig(SqlStatements.class).getAttributes();
        return elSqlBundle.getSql(template, attributes);
    }
}
```

That gets me over the issue of ElSql doing the locating. The attributes will work and I wasn't aware of how the StringTemplateEngine did it until I read the code. Looks like the `@Define` annotation is what I was missing. 

Eventually, I may consider implementing my own annotation that essentially combines the `@Bind` and `@Define` concepts for this kind of integration, but for now, annotating each argument on the Sql object will be good enough.

Thanks for your help in thinking through this.

**Reply by stevenschlansker** [reply]
Great, glad you got it working! If other users are interested, and the support works well, we'd happily accept a contribution for a new elsql module :)
